### ISSUE TITLE
Tree edit operations (NNI,SPR,TBR)

### ISSUE THREAD

**Issue by sriram98v** [sriram98v/phylo-rs#8]

**Comment by swagle8987** [comment]
Hey @sriram98v for the subtree and prune operations, shouldn't they consume the value of the tree sent in instead of borrowing a mutable reference to self? It would allow us to be more explicit in the fact that these operations are changing the structure of the tree and that the tree after the operation is not the same. If we aren't mutating the original tree then we could just pass a reference anyways instead of sending a mutable reference.

**Comment by sriram98v** [comment]
The idea for subtree is to return a copy of the subtree starting at the input node without removing it from self. prune removes the subtree from self (hence using a &mut self) and returns the subtree. Graft consumes the input tree and grafts it to self (mutates self, hence &mut self)

**Comment by swagle8987** [comment]
@sriram98v yes but I think it's better if prune and regraft take ownership of the value rather than borrowing a reference. The subtree method can be a "view" method which returns the subtree at the node(although this wouldnt keep track of the changes made within the subtree of the original tree) but the prune and regraft method fundamentally change the tree and hence explicitly for the end user, they consume the original tree. 
This also avoids bugs since now users don't have to compulsorily keep the tree variable as mutable and can instead create a mutable clone for prune and regraft.

**Comment by jianshu93** [comment]
Hi @swagle8987,

Thanks for the library, in som cases where I need to trim some very large tree according to a list of tips (only keep tips in the list), so for a million taxa, I need to trim 95% of them for downstream calculations, I need to prune it first and collapse the tree so that parent with only one child will be collapsed. I found that many libraries are very slow for this step, the balanced parenthesis approach (for newick format) in C++ (https://github.com/biocore/unifrac-binaries/blob/main/src/tree.cpp) can be extremely fast for this step, is this step also slow for large trees, for example in phylotree-rs (https://github.com/lucblassel/phylotree-rs), according to my test is very slow. 

Best,
Jianshu

**Comment by sriram98v** [comment]
@jianshu93 Did you by chance check Phylo-rs in this regard? I think when you're trying to trim off 95% of the tree, it might be better to use the prune trait with the 5% you want to retain. That should return the tree you want to work with (you will have to drop the other tree here for memory down the line)

**Comment by jianshu93** [comment]
Hi @swagle8987, yes how do I collapse the tree after trimming? since I can imagine there will be many tip trimmed, many branch will only have one child. Do you have some expectation what the complexity for prune operation? Best --Jianshu

**Comment by tkanderson** [comment]
Hi @jianshu93, Please tag @sriram98v - Sriram is the author of this library and if you don't tag him, he may not see your message and be able to reply. Thanks.

**Comment by sriram98v** [comment]
@jianshu93 Your question is not clear to me; can you please clarify the problem you are trying to solve? If you
